core/domain/platform_parameter_domain_test.py

```diff
@@ -81,3 +82,1099 @@ def test_to_dict(self):
+            ))
+        with self.assertRaisesRegexp(
+            utils.ValidationError, 'Unsupported comparison operator \'!=\''):
+            filter_domain.validate()
+
+    def test_validate_filter_with_invalid_server_mode_raises_exception(self):
+        filter_domain = (
+            parameter_domain
+            .PlatformParameterFilter.from_dict(
+                {'type': 'server_mode', 'conditions': [('=', 'invalid')]}
+            ))
+        with self.assertRaisesRegexp(
+            utils.ValidationError, 'Invalid server mode \'invalid\''):
+            filter_domain.validate()
+
+    def test_validate_filter_with_invalid_user_locale_raises_exception(self):
+        filter_domain = (
+            parameter_domain
+            .PlatformParameterFilter.from_dict(
+                {'type': 'user_locale', 'conditions': [('=', 'invalid')]}
+            ))
+        with self.assertRaisesRegexp(
+            utils.ValidationError, 'Invalid user locale \'invalid\''):
+            filter_domain.validate()
+
+    def test_validate_filter_with_invalid_client_type_raises_exception(self):
+        filter_domain = (
+            parameter_domain
+            .PlatformParameterFilter.from_dict(
+                {'type': 'client_type', 'conditions': [('=', 'invalid')]}
+            ))
+        with self.assertRaisesRegexp(
+            utils.ValidationError, 'Invalid client type \'invalid\''):
+            filter_domain.validate()
+
+    def test_validate_filter_with_invalid_version_expr_raises_exception(self):
+        filter_domain = (
+            parameter_domain
+            .PlatformParameterFilter.from_dict(
+                {'type': 'app_version', 'conditions': [('=', '1.a.2')]}
+            ))
+
+        with self.assertRaisesRegexp(
+            utils.ValidationError, 'Invalid version expression \'1.a.2\''):
+            filter_domain.validate()
+
+
+class PlatformParameterRuleTests(test_utils.GenericTestBase):
+    """Test for the PlatformParameterRule."""
+
+    def test_from_dict(self):
+        filters = [
+            {
+                'type': 'app_version',
+                'conditions': [('=', '1.2.3')]
+            },
+            {
+                'type': 'server_mode',
+                'conditions': [('=', 'dev'), ('=', 'test')]
+            }
+        ]
+        rule = parameter_domain.PlatformParameterRule.from_dict(
+            {
+                'filters': filters,
+                'value_when_matched': False,
+            },
+        )
+        self.assertIsInstance(rule, parameter_domain.PlatformParameterRule)
+
+        filter_domain = rule.filters[0]
+        self.assertIsInstance(
+            filter_domain, parameter_domain.PlatformParameterFilter)
+        self.assertEqual(len(rule.filters), 2)
+        self.assertEqual(filter_domain.type, 'app_version')
+        self.assertEqual(filter_domain.conditions, [('=', '1.2.3')])
+        self.assertEqual(rule.value_when_matched, False)
+
+    def test_to_dict(self):
```

Comment:
Here and elsewhere: even these tests should still have clearly defined conditions and outcomes.

Reply:
Done, named them to `test_to_dict_returns_correct_dict` and others accordingly.